src/api/types/mobility.ts

```diff
@@ -46,3 +48,147 @@ export type ViolationsReportQuery = {
+    .int()
+    .optional()
+    .nullable()
+    .describe(
+      'The minute at which the rate will no longer apply (exclusive). If not provided, the rate applies until the trip ends.',
+    ),
+  interval: z
+    .number()
+    .int()
+    .describe(
+      'Interval in minutes at which the rate of this segment is reapplied. An interval of 0 indicates the rate is only charged once.',
+    ),
+  rate: z
+    .number()
+    .describe(
+      'Rate that is charged for each minute interval after the start. Can be a negative number, indicating a discount.',
+    ),
+});
+
+const ShmoPricingPlanSchema = z.object({
+  currency: z.string().describe('Currency in ISO 4217 code'),
+  price: z.number().describe('Fare price in the specified currency'),
+  perMinPricing: z
+    .array(ShmoPricingSegmentSchema)
+    .optional()
+    .nullable()
+    .describe('Array of pricing segments, optional'),
+});
+
+export enum ShmoBookingState {
+  NOT_STARTED = 'NOT_STARTED',
+  IN_USE = 'IN_USE',
+  PAUSED = 'PAUSED',
+  FINISHING = 'FINISHING',
+  FINISHED = 'FINISHED',
+  CANCELLED = 'CANCELLED',
+}
+
+// Inferring schema from enum instead of the other way around
+// this allows the enums to be used directly
+const ShmoBookingStateSchema = z.enum(
+  Object.values(ShmoBookingState) as [ShmoBookingState, ...ShmoBookingState[]],
+);
+
+const ShmoPricingSchema = z.object({
+  currentAmount: z.number(),
+  finalAmount: z.number().optional().nullable(),
+});
+
+const ShmoOperatorSchema = z.object({
+  id: z.string(),
+  name: z.string(),
+});
+
+export const ShmoBookingSchema = z.object({
+  bookingId: z.string().uuid(),
+  state: ShmoBookingStateSchema,
+  orderId: z.string(),
+  pricingPlan: ShmoPricingPlanSchema,
+  departureTime: z.coerce.date().optional().nullable(),
+  arrivalTime: z.coerce.date().optional().nullable(),
+  operator: ShmoOperatorSchema,
+  stateOfCharge: z.number().int().optional().nullable(),
+  currentRangeKm: z.number().int().optional().nullable(),
+  pricing: ShmoPricingSchema,
+});
+
+export type ShmoBooking = z.infer<typeof ShmoBookingSchema>;
+
+const longitude = z
+  .number()
+  .min(-180, {message: 'longitude must be greater than or equal to -180'})
+  .max(180, {message: 'longitude must be less than or equal to 180'});
+const latitude = z
+  .number()
+  .min(-90, {message: 'latitude must be greater than or equal to -90'})
+  .max(90, {message: 'latitude must be less than or equal to 90'});
+
+const ShmoCoordinatesSchema = z.object({
+  longitude,
+  latitude,
+  altitude: z.number().optional().nullable(),
+});
+
+const InitShmoOneStopBookingRequestBodySchema = z.object({
+  recurringPaymentId: z.number().int(),
+  coordinates: ShmoCoordinatesSchema,
+  assetId: z
+    .string()
+    .optional()
+    .describe('This is the same id as vehicleId from the mobility API'),
+});
+
+export type InitShmoOneStopBookingRequestBody = z.infer<
+  typeof InitShmoOneStopBookingRequestBodySchema
+>;
+
+export enum ShmoBookingEventType {
```

Comment:
Interesting to not have an "Unknown", "Init", "Start booking" event here, is that intended or how does it work?

Reply:
Nice observation!

For now we get everything needed to start a trip with the init call, and after that only support ending a trip (which is done in two steps: 1. `START_FINISHING` to end the part you pay for, and 2. `FINISH` to also be able to send a parking photo).

As the TOMP API standard is designed to support all sorts of use cases for shared mobility, there are several more [there](https://app.swaggerhub.com/apis-docs/TOMP-API-WG/transport-operator_maas_provider_api/1.5.0#/trip%20execution/post_legs__id__events): `PREPARE`, `ASSIGN_ASSET`, `SET_IN_USE`, `TIME_EXTEND`, `TIME_POSTPONE`, `PAUSE`, `OPEN_TRUNK`.

Especially `PAUSE` AND `SET_IN_USE` are likely to be used by us in the future to support pausing and resuming a trip.

Reply:
And btw this is already supported in the [Entur API](https://petstore.swagger.io/?url=https://api.dev.entur.io/api-docs/shared-mobility#/Booking/oneStopEvent) with slightly easier to understand event names (AtB doesn't use TOMP directly):
<img width="421" alt="Skjermbilde 2024-09-19 kl  14 35 35" src="https://github.com/user-attachments/assets/15c89d05-4d1a-4ea9-bcaa-3ab18103eb6e">

Reply:
Got it, so there are possibilities to expand the events in the future, after the integration are properly implemented.

Reply:
Indeed 😄